identity-account/src/updates/update.rs

```diff
@@ -0,0 +1,465 @@
+        );
+
+        // The verification method must not exist.
+        ensure!(
+          state
+            .document()
+            .resolve_method(location.fragment().identifier())
+            .is_none(),
+          UpdateError::DuplicateKeyFragment(location.fragment().clone()),
+        );
+
+        let public: PublicKey = if let Some(method_private_key) = method_secret {
+          insert_method_secret(storage, did, &location, type_, method_private_key).await
+        } else {
+          storage.key_new(did, &location).await
+        }?;
+
+        let method: IotaVerificationMethod =
+          IotaVerificationMethod::from_did(did.to_owned(), KeyType::Ed25519, &public, location.fragment().name())?;
+
+        state.store_method_generations(location.fragment().clone());
+
+        // We can ignore the result: we just checked that the method does not exist.
+        state.document_mut().insert_method(method, scope);
+      }
+      Self::DeleteMethod { fragment } => {
+        let fragment: Fragment = Fragment::new(fragment);
+
+        // The verification method must exist
+        ensure!(
+          state.document().resolve_method(fragment.identifier()).is_some(),
+          UpdateError::MethodNotFound
+        );
+
+        // TODO: Do we have to ? here?
+        let method_url: IotaDIDUrl = did.to_url().join(fragment.identifier())?;
+        let core_method_url: CoreDIDUrl = CoreDIDUrl::from(method_url.clone());
+
+        // Prevent deleting the last method capable of signing the DID document.
+        let capability_invocation_set = state.document().as_document().capability_invocation();
+        let is_capability_invocation = capability_invocation_set
+          .iter()
+          .any(|method_ref| method_ref.id() == &core_method_url);
+
+        ensure!(
+          !(is_capability_invocation && capability_invocation_set.len() == 1),
+          UpdateError::InvalidMethodFragment("cannot remove last signing method")
+        );
+
+        state.document_mut().remove_method(method_url).unwrap();
+      }
+      Self::AttachMethod {
+        fragment,
+        relationships,
+      } => {
+        let fragment: Fragment = Fragment::new(fragment);
+
+        // TODO: Do we have to ? here?
+        let method_url: IotaDIDUrl = did.to_url().join(fragment.identifier())?;
+
+        // The verification method must exist
+        ensure!(
+          state.document().resolve_method(fragment.identifier()).is_some(),
+          UpdateError::MethodNotFound
+        );
+
+        // The verification method must not be embedded.
+        ensure!(
+          !state
+            .document()
+            .as_document()
+            .verification_relationships()
+            .any(|method_ref| match method_ref {
+              MethodRef::Embed(method) => method.id().fragment() == method_url.fragment(),
+              MethodRef::Refer(_) => false,
+            }),
+          UpdateError::InvalidTargetMethod
+        );
```

Comment:
I think this was discussed before but is it not possible to attach methods to more than one verification relationship even if it's embedded already?

Reply:
I thought with the agreement on introducing `MethodRelationship` the consensus was that attaching relationships on embedded methods is an error. Otherwise, that method would have to be moved to the verification method set and a reference would have to be inserted in the place where it was previously embedded. That's not terribly complicated, but I don't know if it's expected behavior by users?

Reply:
I'm unsure. I was going by the behaviour of `IotaDocument::attach_method_relationship` which only checks that the method is part of the document (as any relationship or just a verification method) then adds another reference to it, unless I'm reading it wrong? Basically the low-level API now allows attaching multiple relationships to embedded methods, but not the `Account`. (code below)

I'm still unsure if the DID specification restricts that a method with multiple relationships MUST be a `verificationMethod` and not embedded. https://www.w3.org/TR/did-core/#verification-relationships

```rust
// Attaches the given relationship to the given method, if the method exists.
pub fn attach_method_relationship(&mut self, did_url: IotaDIDUrl, relationship: MethodRelationship) -> Result<bool> {
  // Ensure the method exists
  self
    .document
    .resolve_method(did_url.url()) <============= HERE: will return any method, even an embedded one
    .ok_or(identity_did::Error::QueryMethodNotFound)?;
  
  let core_did_url: CoreDIDUrl = CoreDIDUrl::from(did_url);
  let method_ref = MethodRef::Refer(core_did_url);
  
  let result = match relationship {
    MethodRelationship::Authentication => self.document.authentication_mut().append(method_ref),
    MethodRelationship::AssertionMethod => self.document.assertion_method_mut().append(method_ref),
    MethodRelationship::KeyAgreement => self.document.key_agreement_mut().append(method_ref),
    MethodRelationship::CapabilityDelegation => self.document.capability_delegation_mut().append(method_ref),
    MethodRelationship::CapabilityInvocation => self.document.capability_invocation_mut().append(method_ref),
  };

  Ok(result)
}
```

Reply:
Actually, I'm not quite right on the behaviour of `resolve_method`. It depends on the order that the relationships are added. 

E.g. 
1. if a method is embedded as an `Authentication` relationship then `attach`ed as an `AssertionMethod` reference, `resolve_method` will return the embedded method.
2. if a method is embedded as a `CapabilityInvocation` relationship then `attach`ed as an `Authentication` reference, `resolve_method` will return `None`.

This is due to the following code, note the match of `Method::Refer` only looks at `verificationMethod`.

```rust
  fn resolve_method_inner(&self, query: MethodQuery<'_>) -> Option<&VerificationMethod<U>> {
    let mut method: Option<&MethodRef<U>> = None;

    if method.is_none() {
      method = self.authentication.query(query.clone());
    }

    if method.is_none() {
      method = self.assertion_method.query(query.clone());
    }

    if method.is_none() {
      method = self.key_agreement.query(query.clone());
    }

    if method.is_none() {
      method = self.capability_delegation.query(query.clone());
    }

    if method.is_none() {
      method = self.capability_invocation.query(query.clone());
    }

    match method {
      Some(MethodRef::Embed(method)) => Some(method),
      Some(MethodRef::Refer(did)) => self.verification_method.query(&did.to_string()),
      None => self.verification_method.query(query),
    }
  }
```

**Conclusion:** based on this code, you are correct that we should *not* allow references to methods embedded as relationships. My suggestion would be to update `attach_method_relationship` to use `resolve_method_with_scope(method, MethodScope::VerificationMethod)` and return an error like `MethodEmbeddedOrNotFound`. A few new tests might be needed.

The alternative would be to update `resolve_method_inner` to resolve embedded references properly. Let me know if any of my assumptions here are incorrect.

Reply:
Added the `resolve_method_with_scope` check and kept the original error. Will make the errors better in the future, when the `UpdateError::InvalidTargetEmbeddedMethod` moves to the lower-level API.